src/serprog.c

```diff
@@ -0,0 +1,510 @@
+
+enum spi_mode {
+    SPI_MODE_HALF_DUPLEX = 0,
+    SPI_MODE_FULL_DUPLEX = 1,
+    SPI_MODE_MAX = SPI_MODE_FULL_DUPLEX,
+};
+
+enum cs_mode {
+    CS_MODE_AUTO = 0,
+    CS_MODE_SELECTED = 1,
+    CS_MODE_DESELECTED = 2,
+    CS_MODE_MAX = CS_MODE_DESELECTED,
+};
+
+// little endian helper functions
+
+static uint16_t read_le16(const unsigned char *buf) {
+    return buf[0] | (buf[1] << 8);
+}
+
+static uint32_t read_le32(const unsigned char *buf) {
+    return buf[0] | (buf[1] << 8) | (buf[2] << 16) | (buf[3] << 24);
+}
+
+static void write_le24(unsigned char *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+}
+
+static void write_le32(unsigned char *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+    buf[3] = val >> 24;
+}
+
+// serprog communication functions
+
+static int perform_serprog_cmd_full(const PROGRAMMER *pgm, uint8_t cmd,
+                                    const unsigned char *params, int params_len,
+                                    const unsigned char *send_buf, int send_len,
+                                    unsigned char *recv_buf, int recv_len) {
+    unsigned char resp_status_code = 0;
+
+    serial_send(&pgm->fd, &cmd, 1);
+    if (params_len > 0)
+        serial_send(&pgm->fd, params, params_len);
+    if (send_len > 0)
+        serial_send(&pgm->fd, send_buf, send_len);
+
+    if (serial_recv(&pgm->fd, &resp_status_code, 1) < 0 || serial_recv(&pgm->fd, recv_buf, recv_len) < 0)
+        return -1;
+
+    return resp_status_code == S_ACK? 0: resp_status_code == S_NAK? 1: -1;
+}
+
+static int perform_serprog_cmd(const PROGRAMMER *pgm, uint8_t cmd,
+                               const unsigned char *params, int params_len,
+                               unsigned char *recv_buf, int recv_len) {
+    return perform_serprog_cmd_full(pgm, cmd, params, params_len, NULL, 0, recv_buf, recv_len);
+}
+
+/**
+ * @brief Sends/receives a message to the AVR in full duplex mode
+ * @return -1 on failure, otherwise number of bytes sent/received
+ */
+static int serprog_spi_duplex(const PROGRAMMER *pgm, const unsigned char *tx, unsigned char *rx, int len) {
+    unsigned char params[6];
+    write_le24(params, len);
+    write_le24(params + 3, len);
+    if (perform_serprog_cmd_full(pgm, S_CMD_O_SPIOP, params, sizeof(params), tx, len, rx, len) != 0) {
+        return -1;
+    }
+
+    return len;
+}
+
+static bool is_serprog_cmd_supported(const unsigned char *cmd_bitmap, unsigned char cmd) {
+    return (cmd_bitmap[cmd / 8] >> (cmd % 8)) & 1;
+}
+
+// programmer lifecycle handlers
+
+static int serprog_open(PROGRAMMER *pgm, const char *pt) {
+    const char *port_error =
+        "unknown port specification, "
+        "please use the format /dev/ttyACM0\n";
+    char port_default[] = "/dev/ttyACM0";
+    char *serialdev;
+    char *port = mmt_strdup(pt);
+
+    if (str_eq(port, "unknown")) {
```

Comment:
It is uncommon for `-c` programmers to have their own fallbacks on default ports. The AVRDUDE per-user config file allows setting default ports for your system. My `~/.avrduderc` starts with
```
default_serial = "/dev/ttyUSB0";
default_programmer = "urclock";
allow_subshells = yes;
```

I advise against the syntax `-P unknown` with fallback that works for your OS. Even if that was documented (how would users know otherwise), it is out of character for AVRDUDE to offer that.